libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,672 @@
+    // @Description: Number of seconds after completion of tune to auto-save. This is useful when using a 2 position switch for quicktune
+    // @Units: s
+    // @User: Standard
+    AP_GROUPINFO("AUTO_SAVE", 11, AP_Quicktune, auto_save, 0),
+
+    // @Param: MAX_REDUCE
+    // @DisplayName: Quicktune maximum gain reduction
+    // @Description: This controls how much quicktune is allowed to lower gains from the original gains. If the vehicle already has a reasonable tune and is not oscillating then you can set this to zero to prevent gain reductions. The default of 20% is reasonable for most vehicles. Using a maximum gain reduction lowers the chance of an angle P oscillation happening if quicktune gets a false positive oscillation at a low gain, which can result in very low rate gains and a dangerous angle P oscillation.
+    // @Units: %
+    // @Range: 0 100
+    // @User: Standard
+    AP_GROUPINFO("MAX_REDUCE", 12, AP_Quicktune, max_reduce, 20),
+
+    // @Param: OPTIONS
+    // @DisplayName: Quicktune options
+    // @Description: Additional options. When the Two Position Switch option is enabled then a high switch position will start the tune, low will disable the tune. you should also set a QUIK_AUTO_SAVE time so that you will be able to save the tune.
+    // @Bitmask: 0:UseTwoPositionSwitch
+    // @User: Standard
+    AP_GROUPINFO("OPTIONS", 13, AP_Quicktune, options, 0),
+
+    AP_GROUPEND
+};
+
+// Call at loop rate
+void AP_Quicktune::update(){
+
+    if (enable < 1){
+        return;
+    }
+    uint32_t now = AP_HAL::millis();
+
+    if (now - last_update < UPDATE_PERIOD_MS) {
+        return;
+    }
+    last_update = now;
+
+    if (have_pilot_input()){
+        last_pilot_input = now;
+    }
+
+    qt_switch_pos sw_pos_tune = qt_switch_pos::MID;
+    qt_switch_pos sw_pos_save = qt_switch_pos::HIGH;
+    if ((options & OPTIONS_TWO_POSITION) != 0){
+        sw_pos_tune = qt_switch_pos::HIGH;
+        sw_pos_save = qt_switch_pos::NONE;
+    }
+
+    const auto &vehicle = *AP::vehicle();
+
+    if (sw_pos == sw_pos_tune && (!hal.util->get_soft_armed() || !vehicle.get_likely_flying()) && now > last_warning + 5000){
+        GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "Tuning: Must be flying to tune");
+        last_warning = now;
+        return;
+    }
+    if (sw_pos == qt_switch_pos::LOW || !hal.util->get_soft_armed() || !vehicle.get_likely_flying()){
+        // Abort, revert parameters
+        if (need_restore){
+            need_restore = false;
+            restore_all_params();
+            GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "Tuning: Reverted");
+            tune_done_time = 0;
+        }
+        reset_axes_done();
+        return;
+    }
+    if (sw_pos == sw_pos_save){
+        // Save all params
+        if (need_restore){
+            need_restore = false;
+            save_all_params();
+            GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "Tuning: Saved");
+        }
+    }
+    if (sw_pos != sw_pos_tune){
```

Comment:
```suggestion
    if (sw_pos != sw_pos_tune) {
```

and elsewhere; find/replace

Reply:
done